Updates to FreeAssAlgIdeal

Sequenzer opened the pull request:
- Added `deg_bound` field to `FreeAssAlgIdeal` struct for speed improvement
- Modified constructor to initialize `deg_bound` and `gb`.
- Implemented `Base.show` for `FreeAssAlgIdeal`. 
- Added equality and membership functions for `FreeAssAlgIdeal`.
- Updated `ideal` function. 
- Enhanced `groebner_basis` functions with `interreduce` option.
- Added `is_groebner_basis` and `interreduce!` functions.
- Added some more test for integrated testing with `quantum_automorphism_groups`.
- Renamed `FreeAssAlg` -> `FreeAssociativeAlgebra`

Sequenzer committed:
some changes

Sequenzer committed:
added ordering support, as well as minor changes, plus some tests

Sequenzer committed:
Merge branch 'oscar-system:master' into feature/ncgroebner-termorder

joschmitt commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -98,24 +110,26 @@ true
 
-function Base.in(a::FreeAssAlgElem, I::FreeAssAlgIdeal, deg_bound::Int)
-  return ideal_membership(a, I, deg_bound)
+function Base.in(I1, I2::FreeAssAlgIdeal)
```
Maybe this should rather be `is_subset` instead of `in`?

lgoettgens replied:
I think this should be as similar as possible (w.r.t naming an interface) to ideals in MPolyRing. (not sure right now how they do this particular thing though)

Sequenzer committed:
in to is_subset

Sequenzer committed:
fixed the slowdown caused by is_groebner_basis (The function is still slow).

Sequenzer committed:
improving codecov

Sequenzer committed:
even more tests for coverage

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -17,18 +17,24 @@ Two-sided ideal of a free associative algebra with elements of type `T`.
+Base.show(io::IO, a::FreeAssAlgIdeal) = print(io, "Ideal of ", base_ring(a), " with ", number_of_generators(a), " generators")
+
+function Base.:(==)(I2::FreeAssAlgIdeal, I1::FreeAssAlgIdeal)
+  return is_subset(I1,I2) && is_subset(I2,I1)
```
```suggestion
  check_base_rings(I1, I2)
  I1 === I2 && return true
  gens(I1) == gens(I2) && return true
  return is_subset(I1, I2) && is_subset(I2, I1)
```
some optimizations copied from the MPolyIdeal implementation at https://github.com/oscar-system/Oscar.jl/blob/c97f041686553748072fd508e54d0eee53f5d612/src/Rings/mpoly-ideals.jl#L1504-1509

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -98,24 +111,26 @@ true
-function Base.in(a::FreeAssAlgElem, I::FreeAssAlgIdeal, deg_bound::Int)
-  return ideal_membership(a, I, deg_bound)
+function is_subset(I1::FreeAssAlgIdeal, I2::FreeAssAlgIdeal)
+  return all(x -> in(x, I2), gens(I1))
```
```suggestion
  check_base_rings(I1, I2)
  return all(x -> in(x, I2), gens(I1))
```

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -37,16 +43,20 @@ Return the two-sided ideal of $R$ generated by $g$.
+end
 
+function ideal(g::IdealGens{T}, check_groebner_basis::Bool=false) where T <: FreeAssAlgElem
+  return FreeAssAlgIdeal(g, check_groebner_basis)
```
```suggestion
function ideal(g::IdealGens{T}; check_groebner_basis::Bool=false) where T <: FreeAssAlgElem
  return FreeAssAlgIdeal(g; check_groebner_basis)
```
for readability, this should better be a kwarg

Sequenzer replied:
Somehow I didn't see those comments

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -17,18 +17,24 @@ Two-sided ideal of a free associative algebra with elements of type `T`.
-
-  function FreeAssAlgIdeal(R::FreeAssAlgebra, g::Vector{T}) where T <: FreeAssAlgElem
+  deg_bound::Int
+  function FreeAssAlgIdeal(g::IdealGens{T}, check_groebner_basis::Bool=false) where T <: FreeAssAlgElem
```
```suggestion
  function FreeAssAlgIdeal(g::IdealGens{T}; check_groebner_basis::Bool=false) where T <: FreeAssAlgElem
```
see below

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -17,18 +17,24 @@ Two-sided ideal of a free associative algebra with elements of type `T`.
-    r.gens = IdealGens(R, g)
+    r.gens = g
+    @req !check_groebner_basis || base_ring(base_ring(g)) isa Field "Groebner basis check requires a field base ring"
+    if check_groebner_basis && base_ring(base_ring(g)) isa Field && is_groebner_basis(g)
```
I don't really understand what this new arg `check_groebner_basis` is supposed to do?
Could you write some explanation about that or maybe just put that directly into a docstring for `ideal(::IdeaGens; check_groebner_basis::Bool)`?

Sequenzer replied:
It is supposed to check if the generator is already a Groebner basis and then add it to the .gb immediately. But in retrospect, I do not think it helps that much, especially if it is hidden behind a kwarg. I remove it.

lgoettgens reviewed (commented):
some misc comments of things that confused me

lgoettgens commented on src/Rings/FreeAssAlgIdeal.jl:
```diff
@@ -17,18 +17,24 @@ Two-sided ideal of a free associative algebra with elements of type `T`.
-  return Expr(:call, :ideal, [expressify(g, context = context) for g in collect(a.gens)]...)
+Base.show(io::IO, a::FreeAssAlgIdeal) = print(io, "Ideal of ", base_ring(a), " with ", number_of_generators(a), " generators")
+
+function Base.:(==)(I2::FreeAssAlgIdeal, I1::FreeAssAlgIdeal)
```
every type that implements `==` must also implement `hash(::T, ::UInt)` to allow for correctness with julia Base (see e.g. https://github.com/oscar-system/Oscar.jl/issues/2222).
As there is no normal form for these ideals, one needs to use the limited information one has to construct a really bad (but valid) hash function, e.g. as
```julia
function Base.hash(I::FreeAssAlgIdeal, h::UInt)
  return hash(base_ring(I), h)
end
```

Sequenzer committed:
added Lars suggestion for hash function

Sequenzer committed:
`FreeAssAlgIdeal` -> `FreeAssociativeAlgebraIdeal`

Sequenzer committed:
update AbstractAlgebra and rename FreeAssAlgebra

- Updated AbstractAlgebra version from 0.42.0 to 0.42.3 in Project.toml
- `FreeAssAlgebra` -> `FreeAssociativeAlgebra`

Sequenzer committed:
`FreeAssAlgElem` -> `FreeAssociativeAlgebraElem`

lgoettgens commented on src/Serialization/Algebras.jl:
```diff
@@ -1,28 +1,28 @@
 
 # Free associative algebra serialization
-@register_serialization_type FreeAssAlgebra uses_id
+@register_serialization_type FreeAssociativeAlgebra uses_id
```
@antonydellavecchia Iirc this renaming here and the two further ones in lines 21 and 56 in this file need some mention in a serialization upgrade script. Do we already have some code snippet to put in there for such a simple type renaming or, if not, could you put together something for it?

Sequenzer replied:
I believe @antonydellavecchia will write an update script...

antonydellavecchia replied:
including something like this in src/Serialization/Upgrades
and naming the file 1.2.0.jl, I copied this script from 0.15.0.jl,
or wtv the next version of oscar will be I guess 1.2.0

```
push!(upgrade_scripts_set, UpgradeScript(
  v"1.2.0",
  function upgrade_1_2_0(s::UpgradeState, dict::Dict)

    upgraded_dict = dict

    renamings = Dict{String,String}([
      ("FreeAssAlgebra", "FreeAssociativeAlgebra"),
      ("FreeAssAlgElem", "FreeAssociativeAlgebraElem"),
      ("FreeAssAlgIdeal", "FreeAssociativeAlgebraIdeal"),
    ])

    function upgrade_type(d::String)
      return get(renamings, d, d)
    end
    function upgrade_type(d::Dict)
      upg_d = d
      upg_d[:name] = get(renamings, d[:name], d[:name])
      if d[:params] isa Dict && haskey(d[:params], :_type)
        upg_d[:params][:_type] = upgrade_type(d[:params][:_type])
      elseif d[:params] isa Vector
        upg_d[:params] = [upgrade_type(v) for v in d[:params]]
      end
      return upg_d
    end

    if haskey(dict, :_type)
      upgraded_dict[:_type] = upgrade_type(dict[:_type])
    end
        
    if haskey(dict, :data) && dict[:data] isa Dict
      upgraded_dict[:data] = upgrade_1_2_0(s, dict[:data])
    end

    if haskey(dict, :_refs)
      upgraded_refs = Dict()
      for (k, v) in dict[:_refs]
        upgraded_refs[k] = upgrade_1_2_0(s, v)
      end
      upgraded_dict[:_refs] = upgraded_refs
    end

    return upgraded_dict
  end
))
```

lgoettgens replied:
Great! Thanks @antonydellavecchia. Some remarks:
- we are only in the 1.2.0 cycle, so all 1.3.0 should be changed to 1.2.0.
- near the bottom are still two old 0_15_0, should be 1_2_0 as well
- the renamings list is missing the ideal

@Sequenzer can you please incorporate this script with the changes lined out here? I will try it out once we have this in this PR.

Edit: I changed the above code comment and included the changes there directly

lgoettgens replied:
I verified that this upgrade script works when trying to load some simple objects that were saved with Oscar 1.1.1. (For things that were saved with 1.2.0-DEV versions the upgrade script does not run, but this is not relevant for this PR)

Sequenzer committed:
upgrade script

Sequenzer committed:
fix of kwarg change in FreeAssAlgebraIdeal

Sequenzer committed:
removed the check_groebner_basis functionality

Sequenzer committed:
switched out assert, added a base_ring check to is_subset

antonydellavecchia committed:
refactored upgrade types + fix upgrade script for dicts + test

lgoettgens commented:
@antonydellavecchia It seems that [`8990f92` (#4035)](https://github.com/oscar-system/Oscar.jl/pull/4035/commits/8990f925154bba28bf26ddd2c7a63b168a3cd1f1) broke CI

antonydellavecchia committed:
fix tests

antonydellavecchia committed:
fixes backwards compatibility test

lgoettgens committed:
Merge branch 'master' into feature/ncgroebner-termorder

fingolfin commented:
I'll take a look at this PR this afternoon, and either merge it or request changes. If I haven't done it by the end of the day, feel free to just merge, I can always get changes in later as well.

lgoettgens reviewed (approved):
All comments have been addressed. And Max didn't do a review yesterday, so let's merge this for now.

lgoettgens committed:
Merge branch 'master' into feature/ncgroebner-termorder
